### ISSUE TITLE
Pysat returns 0 or Error

### ISSUE THREAD

**Issue by stijndebels** [CPMpy/cpmpy#295]
Hello,

As I'm working with cpmpy for my thesis, I stumbled upon an issue using Pysat. When I try for frequent itemset mining in Pysat, I get the following error: 
`NotImplementedError: CPM_pysat: Non supported constraint (~transactions[0]) -> (sum([1, 0, 1, 1, 1] * [items[0], items[1], items[2], items[3], items[4]]) > 0)
`
This belongs to the  coverage constraint `m += [Trans[t] == (cpmpy.sum([(1 - TDB[t][i]) * Items[i] for i in range(nrI)]) <= 0) for t in range(nrT)]`

When I then rewrite it to this form it works:
`m += [Trans[t] == ([((TDB[t][i]==1) | (Items[i]==0)) for i in range(nrI)] == 0) for t in range(nrT)]`

This works, I get the correct amount of items (31). But I want to find frequent items so I add the constraint:
`m += ((Trans[:].sum()) >= freq)
`
Now I get 0 items, which is not correct. This is weird cause in a SAT paper they have this exact same constraint. It also works with the other solvers. This program should also work for pysat since we are working with only boolvars. In attachment I'll put a small py file so you can check this also.

[pysatTest.py.zip](https://github.com/CPMpy/cpmpy/files/11470218/pysatTest.py.zip)

It would be nice to include the Pysat solver in my thesis, so if anyone knows what's wrong or if something is not fully correct within the pysat conversion, please let me know. It would be nice if this could get fixed.

Kind regards
Stijn

**Comment by tias** [comment]
Hi Stijn,

Thanks for the example.

If I run your test and print the model, I get:
```
Constraints:
    [~transactions[0], ~transactions[1], ~transactions[2], ~transactions[3], ~transactions[4], ~transactions[5], ~transactions[6], ~transactions[7], ~transactions[8], ~transactions[9]]
    sum([transactions[0], transactions[1], transactions[2], transactions[3], transactions[4], transactions[5], transactions[6], transactions[7], transactions[8], transactions[9]]) >= 2
    sum([items[0], items[1], items[2], items[3], items[4]]) >= 1
Objective: None
0
```
so there are no coverage constraints actually posted, so there must be a bug in your rewriting...

**Comment by stijndebels** [comment]
Dear Professor Guns,

You are correct! There was a little thing I forgot to add, it works now (without reification). But my reification constraints still contain elements that are not supported. I'll sent it updated.
[pysatTest.py.zip](https://github.com/CPMpy/cpmpy/files/11470596/pysatTest.py.zip)

Kind regards
Stijn

**Comment by tias** [comment]
Yes, I also noticed reified sum (for the more fine-grained frequency constraint) isnot supported, I filed a separate minimal bug for it (#296 )

As to your model, here is a cleaned-up version of it that is, cleaner.
```
import numpy as np
from cpmpy import *

TDB = np.array(
      [[False, True, False, False, False],
       [False, False, False, False, True],
       [True, False, True, False, False],
       [True, False, False, False, True],
       [False, True, True, False, False],
       [False, False, False, True, True],
       [False, False, True, True, True],
       [True, True, True, False, False],
       [True, True, False, False, True],
       [True, True, True, False, True]])
nrT, nrI = TDB.shape
freq=2

Items = boolvar(shape=nrI, name="items")
Trans = boolvar(shape=nrT, name="transactions")

m = Model()
#m += [Trans[t] == (~any(Items[i] for i in range(nrI) if not TDB[t,i]))
#      for t in range(nrT)]
# same as above but with numpy magic
for T, negrow in zip(Trans, ~TDB):
    m += T == ~any(Items[np.where(negrow)])

m += (sum(Trans) >= freq)
# The potentially more effective version of the above:
# That this does not work is a bug in the CPMpy PySAT translation
#for I, col in zip(Items, TDB.T):
#    m += I.implies(sum(Trans[np.where(col)] >= freq))

print(m)

stdCounter=m.solveAll(solver='pysat', display=lambda: print(Items.value(), Trans.value()))
print(stdCounter)
```

(no need to attach zip files, you can just copy-paste code)

I guess your real bug is the other one, if so then you can close this issue

**Comment by tias** [comment]
note that you can already run experiments with the above, that works perfectly find on pysat...

**Comment by tias** [comment]
fixed with merge of #316, in master
